README.md
# svKomodo

<!-- badges: start -->

[![R-CMD-check](https://github.com/SciViews/svKomodo/workflows/R-CMD-check/badge.svg)](https://github.com/SciViews/svKomodo/actions) [![Win Build Status](https://ci.appveyor.com/api/projects/status/github/SciViews/svKomodo?branch=master&svg=true)](https://ci.appveyor.com/project/phgrosjean/svKomodo) [![Coverage Status](https://img.shields.io/codecov/c/github/SciViews/svKomodo/master.svg)](https://codecov.io/github/SciViews/svKomodo?branch=master) [![CRAN status](https://www.r-pkg.org/badges/version/svKomodo)](https://cran.r-project.org/package=svKomodo) [![License](https://img.shields.io/badge/license-GPL-blue.svg)](https://www.gnu.org/licenses/gpl-2.0.html) [![Lifecycle: superseded](https://img.shields.io/badge/lifecycle-superseded.svg)](https://www.tidyverse.org/lifecycle/#superseded)

<!-- badges: end -->

R-side code to implement an R editor and IDE in Komodo IDE (<https://www.activestate.com/products/komodo-ide/>) with the SciViews-K extension.

**Note that this package is maintained minimally in order to keep code that may be useful in another context. However, RStudio and VScode with R extensions are the R editors/IDEs that you should use today. Komodo + SciViews-K/svKomodo aimed to develop a similar R IDE back in the 2000's/2010's, but its development was stopped in 2016 when RStudio gained its popularity.**

## Installation

You can install the released version of {svKomodo} from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("svKomodo")
```

You can also install the latest development version. Make sure you have the {remotes} R package installed:

``` r
install.packages("remotes")
```

Use `install_github()` to install the {svKomodo} package from GitHub (source from **master** branch will be recompiled on your machine):

``` r
remotes::install_github("SciViews/svKomodo")
```

R should install all required dependencies automatically, and then it should compile and install {svKomodo}.

## Usage

You can get further help about this package this way: Make the {svKomodo} package available in your R session:

``` r
library("svKomodo")
```

Get help about this package:

``` r
library(help = "svKomodo")
help("svKomodo-package")
```

For further instructions, please, refer to the help pages at <https://www.sciviews.org/svKomodo/>.

## Code of Conduct

Please note that the svKomodo project is released with a [Contributor Code of Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html). By contributing to this project, you agree to abide by its terms.


NEWS.md
# svKomodo 1.0.0

-   README.md file added.

-   NEWS file reworked to use the Markdown format.

-   Man pages are rewritten in Roxygen2.

# svKomodo 0.9-63

-   Starting from R revision \>= 67550, the HTML help port is now retrieved using `tools::startDynamicHelp(NA)`.

# svKomodo 0.9-62

-   Author field rebuild in DESCRIPTION file

-   `.onAttach()` reworked to eliminate `:::` (not allowed by CRAN)

-   `utils::rc.settings` are adjusted in `.onAttach()` and restored in `.onUnload()`

# svKomodo 0.9-61

-   Simplification of internal code to get Komodo location. It does not require a specific version of 'locate' on Unix/Linux any more.

# svKomodo 0.9-60

-   `.onAttach()` is reworked for not issuing warning messages in Mac OS X.

# svKomodo 0.9-59

-   R is now configured to be used with SciViews Komodo (Komodo Edit + SciViews-K) when the package is attached. It was previously configured when it was loaded.

-   Temporary objects are now stored in `SciViews:TempEnv` instead of `TempEnv` and consequently, svMisc \>= 0.9-68 is needed.

-   Further protection before looking for komodo with locate: (1) if locate is not there, do not try to use it, and (2) do not use it on Mac OS X.

# svKomodo 0.9-58

-   The package was not loading correctly when not on `.libPaths`. Fixed.

# svKomodo 0.9-57

-   svKomodo creates now at loading a `.Last.sys()` function in `SciViews:TempEnv` indicating to Komodo Edit that R has quit. This is required to update R menus there.

-   The SciViews-K configuration file is not saved from here anymore (CRAN now does not accept packages that write files elsewhere than in R temp dir). This configuration file is now created by the SciViews-K Komodo plugin, in svStart.R.

# svKomodo 0.9-56

-   NEWS file reworked to use the new Rd format.

# svKomodo 0.9-55

-   On some machines, `system("locate Komodo ...")` fails. Added some more checking in this situation.

-   In `.onLoad()`, slight reworking of the `svPager()` function to avoid calling `.Internal(file.show(....))`.

# svKomodo 0.9-54

-   Startup code is changed... a part of the code in svStart.R in the SciViews-K plugin is now executed here!

-   Save and load SciViews config in `~/.SciViewsConfig.RData` file, so that it is possible to reconfigure R for SciViews just by reloading the present package.

# svKomodo 0.9-53

-   `koCmd()` can now use either socket or file for communication with Komodo.

# svKomodo 0.9-52

-   The package is created from svGUI 0.9-51 (functions that concern interaction with Komodo Edit/IDE). `guiInstall()`, `guiUninstall()`, `guiRefresh()` & `guiAutoRefresh()` are also renamed `koXXX()`, in order to better match their purpose (to communicate with Komodo Edit/IDE only), while the previous implementation was intended for more general purpose. Since, it was never used elsewhere, narrowing its scope allows to simplify the code greatly!

-   The callback mechanism to run commands from a client is now moved from svSocket package to here, so that it can also be used in the svHttp package.

-   `koRefresh()` now looks at the "changed" attribute of the object returned by `objList()`, and so, can make the difference between "no changes", and "changes, but the environment go no more objects".

-   Here under is a transcript of svGUI before functions are moved to svKomodo:

-   `koCmd()` now should prepend \<\<<js>\>\> to the JavaScript code to get it evaluated in Komodo (starting with SciViews-K 0.9-18). Komodo now also accepts RJsonP strings, prepended with \<\<<rjson>\>\>. If there is no code prepended to the string send to Komodo, it is just printed in the local R console. A new 'type' argument specifies what kind of string we send to Komodo. [in svGUI 0.9-48]

-   Use of `svTaskCallbackManager()` of svSocket \>= 0.9-48 to register task callback that are also executed after each R code send by socket clients. [in svGUI 0.9-46]

-   `guiRefresh()` now clears active items and MRU lists in Komodo for non-defined active data frames and lm objects. [in svGUI 0.9-46]

-   Added `guiRefresh()` and `guiAutoRefresh()` to refresh automatically the content of the GUI (Komodo) object explorer and the lists of active objects. [in svGUI 0.9-45]

-   Preparation for CRAN submission: polishing the package. [in svGUI 0.9-44]

-   Made compatible with R 2.6.x (previous version was compatible with R 2.7.0). [in svGUI 0.9-43]

-   `koCmd()` is now more robust and do not issue a warning or an error if the Komodo server is not available (but the error message is returned by the function with a 'try-error' class, so that it can be processed by the caller). [in svGUI 0.9-42]

-   Correction of a bug in the first example of `koCmd()`. [in svGUI 0.9-41]

-   `guiInstall()` now creates a hook to `koCmd()`: `.koCmd()` in `SciViews:TempEnv`. [in svGUI 0.9-41]

-   First version distributed on R-forge. It is completely refactored from older versions (on CRAN since 2003) to make it run with SciViews-K and Komodo Edit (Tinn-R is also supported, but not SciViews-R Console any more). [in svGUI 0.9-40]


R/koCmd.R
#' Connect to the SciViews-K (Komodo Edit/IDE) socket server and run JavaScript code in Komodo
#'
#' If Komodo Edit/IDE with the SciViews-K extension is running on your machine,
#' you can connect to its socket server and run javascript code in it with this
#' function.
#'
#' @param cmd the JavaScript you want to execute in Komodo Edit, in a character
#  string vector.
#' @param data if a named list, replace `<<<name>>>` in cmd for each name of the
#' list by its component first. If a character string, replace `<<<data>>>` in
#' cmd. If `NULL` (by default), do nothing to cmd before submitting it. See the
#' last examples for using data.
#' @param async not used yet!
#' @param host the host where Komodo is located. Currently, only `localhost` is
#' accepted. Can be changed by setting `options(ko.host = ....)`.
#' @param port the socket port where the SciViews-K server is listening, by
#' default, it is port 7052. Can be changed by setting
#' `options(ko.port = ....)`.
#' @param kotype the type of Komodo server in use. Currently
#' (SciViews-K >= 0.9-25), it can be either `"socket"` or `"file"`.
#' @param timeout number of seconds to wait for a response.
#' @param type which type of Komodo command do we send? If `type = "js"` (by
#' default), `cmd` is considered to be JavaScript code to be evaluated in
#' Komodo. If `type = "rjsonp"`, `cmd` is parsed as RJson object with padding
#' (included in a JavaScript function call) and will be evaluated as such in
#' Komodo. if `type = "output"`, the string in `cmd` is considered to be some R
#' output and will be displayed in the Komodo local R console (no evaluation).
#' @param pad a string naming the JavaScript function to run in Komodo, with the
#' constructed Rjson object as argument. If `NULL` (by default), the Rjson
#' object is evaluated directly without padding.
#' @param ... further arguments to pass to `toRjson()`.
#'
#' @return Returns the results of the evaluation of the javascript code in
#' Komodo Edit/IDE if `async = FALSE`. Note that `async = TRUE` is not
#' supported yet.
#'
#' If there is an error, or `cmd` is an invalid JavaScript code, a character
#' string containing javascript error message is returned (this is changed from
#' version 0.9-47, previously a 'try-error' was returned).
#'
#' @details
#' Komodo Edit (https://www.activestate.com/products/komodo-ide/) is an Open
#' Source (MPL, GPL & LGPL) editor based on the excellent Mozilla platform and
#' the powerful Scintilla text editor widget. It runs on many Linux
#' distributions, on Windows and on MacOS. Komodo IDE was a commercial
#' equivalent, but with many tools for developers, especially targeting
#' languages like Perl, Tcl, Python, Ruby, etc. This product is now freely
#' distributed as well and Komodo Edit was deprecated. However, the project does
#' not appear to be actively maintained any more and it may not work on more
#' recent MacOS or Windows versions.
#'
#' `koCmd()` can only talk to Komdo if the SciViews-K socket server is
#' installed. This server is contained in the SciViews-K extension that you can
#' download from https://github.com/SciViews/sciviewsk. See Komodo documentation
#' to know how to install this extension (drag and drop of the extension on the
#' Komodo window works in most platforms).
#'
#' @note
#' Because of security concerns, the SciViews-K server only allows connections
#' from local clients (running on the same computer). This limitation would be
#' relatively easy to eliminate, but at your own risks!
#'
#' Data are returned from Komodo to R by using the JavaScript function
#' `sv.socket.serverWrite()`, see the examples bellow.
#'
#' @seealso [svSocket::startSocketServer()], [svSocket::processSocket()]
#' @export
#' @keywords IO
#' @concept interprocess commnunication Komodo
#'
#' @examples
#' \dontrun{
#' # Make sure you have started Komodo Edit or IDE with the SciViews-K extension
#' # installed on the same machine, and the socket server started and then...
#'
#' # Send JavaScript commands to Komodo
#' # Alert box in Komodo, and then reply to R
#' koCmd(c('alert("Hello from R!");',
#'   sv.socket.serverWrite("Hello from OpenKomodo (" + ko.interpolate.currentFilePath() + ")");'))
#'
#' # Open a web page wih Komodo configuration
#' koCmd("ko.open.URI('about:config','browser');")
#'
#' # Get info from Komodo
#' koCmd("sv.socket.serverWrite(ko.logging.getStack());")
#'
#' # Passing a large amount of data to Komodo, and then, back to R
#' koCmd(paste('sv.socket.serverWrite("', rep(paste(iris, collapse = "\\\\n"), 10), '");', sep = ""))
#'
#' # It is easier to use 'data =' instead of paste() for constructing the JS command
#' koCmd('alert("<<<data>>>");', data = search())
#'
#' # Using a named list for data to replace in the cmd
#' koCmd('alert("This is R version <<<major>>>.<<<minor>>>");', R.version)
#'
#' # Sending incorrect javascript instruction
#' koCmd('nonexistingJSfunction();')
#' # Should return something like:
#' # "ReferenceError: nonexistingJSfunction is not defined"
#'
#' # Sending RjsonP (Rjson with padding) instruction to Komodo
#' koCmd("Hello with RjsonP!", type = "rjsonp", pad = "alert")
#'
#' # This is more useful to pass complex R objects to Komodo
#' koCmd(head(iris), type = "rjsonp", pad = "sv.socket.serverWrite")
#'
#' # Send simple text (no evaluation) to the Komodo R console
#' koCmd("Hello again from R!", type = "output")
#' }
koCmd <- function(cmd, data = NULL, async = FALSE, host = getOption("ko.host"),
port = getOption("ko.port"), kotype = getOption("ko.kotype"), timeout = 2,
type = c("js", "rjsonp", "output"), pad = NULL, ...) {

  type <- match.arg(type)
  if (is.null(host))
    host <- "localhost"  # Default value
  if (is.null(port))
    port <- 7052         # Idem
  if (is.null(kotype))
    kotype <- "file"	# Idem
  cmd <- gsub("\n", "\\\\n", cmd)
  cmd <- paste(cmd, collapse = " ")
  if (is.na(cmd) || is.null(cmd) || length(cmd) == 0) {
    warning("No command supplied in cmd argument")
    return("")
  }
  # Do we need to paste data in the command?
  if (!is.null(data)) {
    "rework" <- function(data) {
      data <- as.character(data)
      data <- gsub("\n", "\\\\\\\\n", data)
      data <- paste(data, collapse = "\\\\n")
      return(data)
    }

    n <- names(data)
    if (is.null(n)) {
      # We assume that we replace '<<<data>>>'
      cmd <- gsub("<<<data>>>", rework(data), cmd)
    } else {# Named data
      # We replace each <<<name>>> in turn
      for (i in 1:length(n))
        cmd <- gsub(paste("<<<", n[i], ">>>", sep = ""),
          rework(data[[n[i]]]), cmd)
    }
  }
  # What type of data do we send?
  cmd <- switch(type,
    js = paste("<<<js>>>", cmd, sep = ""),
    rjsonp = paste("<<<rjsonp>>>", pad, "(",
      paste(toRjson(cmd, ...), collapse = " "), ")", sep = ""),
    cmd)

  otimeout <- getOption("timeout")
  options(timeout = timeout)  # Default timeout is 120 seconds
  # Do we use file or socket server?
  if (kotype == "file") {
    # File is .sv<port> in the temporary dir (/tmp for Linux and Mac OS X)
    tempdir <- "/tmp"
    if (.Platform$OS.type == "windows")
      tempdir <- dirname(tempdir())
    svfile <- file.path(tempdir, paste(".sv", port, sep = ""))
    svoutfile <- paste(svfile, "out", sep = ".")
    # Make sure the output file is deleted
    unlink(svoutfile)
    # Send data to Komodo
    # Need a time-of-day fingerprint in ms
    fp <- as.integer((as.numeric(Sys.time()) %% 24*60*60) * 1000)
    fp <- gsub(" ", "0", format(fp, width = "8", scientific = FALSE))
    cmd <- paste("<<<", fp, ">>>", cmd, sep = "")
    svcon <- file(svfile, open = "w", blocking = TRUE, encoding = "UTF-8")
    try(cat(paste(cmd, collapse = "\n"), file = svcon), silent = TRUE)
    try(close(svcon), silent = TRUE)
    # Wait for output file... or max timeout (add 500ms to this timeout)
    timeout <- Sys.time() + timeout + 0.5
    while (!file.exists(svoutfile) && Sys.time() < timeout)
      Sys.sleep(0.2) # To avoid using too much resources
    if (file.exists(svoutfile)) {
      res <- try(readLines(svoutfile), silent = TRUE)
      Encoding(res) <- "UTF-8"
    } else {
      res <- character(0)
    }
  } else {# This must be a socket server
    tryCatch({
      con <- socketConnection(host = host, port = port, blocking = TRUE)
      writeLines(cmd, con)
      res <- readLines(con)
      close(con)
    }, warning = function(e) {
      stop(simpleError("Komodo socket server is not available!",
      quote(koCmd)))
    })
#    ret <- try(writeLines(cmd, con), silent = TRUE)
#    if (!inherits(ret, "try-error"))
#      res <- try(readLines(con), silent = TRUE)
#    try(close(con), silent = TRUE)
  }
  options(timeout = otimeout)
  return(res)
}


R/svKomodo-internal.R
.loadSvOptions <- function() {
  # Try loading SciViews options from ~/.SciViewsConfig.RData
  argOption <- function(arg) {
    args <- commandArgs(trailingOnly = TRUE)
    regexp <- paste("^", arg, "=", sep = "")
    where <- grepl(regexp, args)
    if (any(where)) {
      arg <- rev(args[where])[1]
      sub(regexp, "", arg)
    } else NULL
  }

  # For those items not defined yet
  .SciViewsConfig <- list() # Default value
  if (file.exists("~/.SciViewsConfig.RData"))
    load("~/.SciViewsConfig.RData")
  # Change options...
  ko.type <- argOption("ko.type")
  if (!is.null(ko.type))
    options(ko.type = ko.type)
  if (is.null(getOption("ko.type")))
    options(ko.type = .SciViewsConfig$ko.type)
  if (is.null(getOption("ko.type")))
    options(ko.type = "http")

  ko.host <- argOption("ko.host")
  if (!is.null(ko.host))
    options(ko.host = ko.host)
  if (is.null(getOption("ko.host")))
    options(ko.host = .SciViewsConfig$ko.host)
  if (is.null(getOption("ko.host")))
    options(ko.host = "localhost")

  ko.port <- argOption("ko.port")
  if (!is.null(ko.port))
    options(ko.port = ko.port)
  if (is.null(getOption("ko.port")))
    options(ko.port = .SciViewsConfig$ko.port)
  if (is.null(getOption("ko.port")))
    options(ko.port = 7052)

  ko.kotype <- argOption("ko.kotype")
  if (!is.null(ko.kotype))
    options(ko.kotype = ko.kotype)
  if (is.null(getOption("ko.kotype")))
    options(ko.kotype = .SciViewsConfig$ko.kotype)
  if (is.null(getOption("ko.kotype")))
    options(ko.kotype = "socket")

  ko.serve <- argOption("ko.serve")
  if (!is.null(ko.serve))
    options(ko.serve = ko.serve)
  if (is.null(getOption("ko.serve")))
    options(ko.serve = .SciViewsConfig$ko.serve)
  if (is.null(getOption("ko.serve")))
    options(ko.serve = 8888)

  ko.activate <- argOption("ko.activate")
  if (!is.null(ko.activate))
    options(ko.activate = ko.activate)
  if (is.null(getOption("ko.activate")))
    options(ko.activate = .SciViewsConfig$ko.activate)
  if (is.null(getOption("ko.activate")))
    options(ko.activate = 8888)

  ko.id <- argOption("ko.id")
  if (!is.null(ko.id))
    options(ko.id = ko.id)
  if (is.null(getOption("ko.id")))
    options(ko.id = .SciViewsConfig$ko.id)
  if (is.null(getOption("ko.id")))
    options(ko.id = "SciViewsK")

  R.id <- argOption("R.id")
  if (!is.null(R.id))
    options(R.id = R.id)
  if (is.null(getOption("R.id")))
    options(R.id = .SciViewsConfig$R.id)
  if (is.null(getOption("R.id")))
    options(R.id = "R")

  R.initdir <- argOption("R.initdir")
  if (!is.null(R.initdir))
    options(R.initdir = R.initdir)
  if (is.null(getOption("R.initdir")))
    options(R.initdir = .SciViewsConfig$R.initdir)
  R.initdir <- getOption("R.initdir")
  if (is.null(R.initdir)) {
    options(R.initdir = NULL)
  } else {
    # Check that the dir exists...
    isfile <- try(file.exists(R.initdir), silent = TRUE)
    if (inherits(isfile, "try-error") || !isfile || !file.info(R.initdir)$isdir)
      options(R.initdir = NULL)
  }

  width <- argOption("width")
  if (!is.null(width))
    options(width = width)
  if (is.null(getOption("width")))
    options(width = .SciViewsConfig$width)
  if (is.null(getOption("width")))
    options(width = 80)

  OutDec <- argOption("OutDec")
  if (!is.null(OutDec))
    options(OutDec = OutDec)
  if (is.null(getOption("OutDec")))
    options(OutDec = .SciViewsConfig$OutDec)
  if (is.null(getOption("OutDec")))
    options(OutDec = ".")

  OutSep <- argOption("OutSep")
  if (!is.null(OutSep))
    options(OutSep = OutSep)
  if (is.null(getOption("OutSep")))
    options(OutSep = .SciViewsConfig$OutSep)
  if (is.null(getOption("OutSep")))
    options(OutSep = ",")
}

.onAttach <- function(lib, pkg) {
  # Make sure config is OK
  .loadSvOptions()

  # Check that usually inactivated rc.settings are set
  assignTemp(".old.rc.settings", rc.settings())
  rc.settings(ipck = TRUE)

  # Create our SciViews task callback manager
  # Should eliminate this???
  ## PhG: inactivated for now, because it makes problems in R!!!
  #assignTemp(".svTaskCallbackManager", svTaskCallbackManager())
  koInstall()

  # Temporary change of a few R options that make problem with SciViews
  # Do not use fancy quotes
  assignTemp(".useFancyQuotes", getOption("useFancyQuotes"))
  options(useFancyQuotes = FALSE)
  # Limit output to 999 items
  assignTemp(".max.print", getOption("max.print"))
  options(max.print = 999)

	  # If svStart function exists in.GlobalEnv, delegate the rest to it!
  if (exists("svStart", envir = .GlobalEnv, mode = "function",
    inherits = FALSE))
    return()

  # Determine which server to start
  type <- getOption("ko.type")
  req <- require
  if (type == "http") {
    req("svHttp", character.only = TRUE, lib.loc = lib)
    if (interactive())
      try(svHttp::start_http_server())
  } else {
    req("svSocket", character.only = TRUE, lib.loc = lib)
    if (interactive())
      try(svSocket::startSocketServer())
  }

  # This comes from svStart... and should really be placed here indeed!
  # Look if and where komodo is installed
  if (Sys.getenv("koAppFile") != "") {
    Komodo <- Sys.getenv("koAppFile")
  } else {
    Komodo <- ""
  }

  if (.Platform$OS.type == "unix") {
    if (Komodo == "")
      Komodo <- "/usr/local/bin/komodo"  # Default location
    if (!file.exists(Komodo))
      Komodo <- Sys.which("komodo")[1]
    if (length(Komodo) == 0 || Komodo == "")
      Komodo <- NULL

    # Just to avoid warnings while compiling outside of Windows...
    readRegistry <- function() return()
  } else {# Windows
    # If komodo path was not passed in environment
    if (Komodo == "") {
      Komodo <- NULL

      err.null <- function(e) return(NULL)
      # On Windows, 'komodo' should be enough
      # But for reasons that escape me, Komodo seems to strip off its
      # own directory from the path variable. So, I have to restore
      # it from the Windows registry :-(

      # Try several ways to get Komodo path from registry.
      key <- "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\App Paths\\komodo.exe"
      Komodo <- tryCatch(readRegistry(key, hive = "HLM")[["(Default)"]],
        error = err.null)

      if (is.null(Komodo) || !file.exists(Komodo)) {
        key <- "Applications\\komodo.exe\\shell\\open\\command"
        Komodo <- tryCatch(readRegistry(key, hive = "HCR")[["(Default)"]],
          error = err.null)
        if (!is.null(Komodo))
          Komodo <- sub(" *\\\"%[1-9\\*].*$", "", Komodo)
      }

      if (is.null(Komodo) || !file.exists(Komodo)) {
        key <- "SYSTEM\\CurrentControlSet\\Control\\Session Manager\\Environment"
        Path <- tryCatch(readRegistry(key, hive = "HLM")$Path,
          error = err.null)
        if (!is.null(Path) && !is.na(Path) && Path != "") {
          Path <- strsplit(Path, ";")[[1]]
          Path <- Path[sapply(Path, function(x)
            file.exists(file.path(x, "komodo.exe")))][1]
          Komodo <- gsub("\\\\+", "\\\\", file.path(Path, "komodo.exe",
            fsep = "\\"))
        }
      }
    }
  }

  if (length(Komodo) && Komodo != "" && file.exists(Komodo)) {
    # Change the editor and the pager to Komodo
    options(pager2 = getOption("pager"))
    # A custom pager consists in displaying the file in Komodo
    svPager <- function(files, header, title, delete.file) {
      files <- gsub("\\", "\\\\", files[1], fixed = TRUE)
      res <- tryCatch(koCmd(sprintf('sv.r.pager("%s", "%s")', files, title)),
        error = function(e) return(FALSE))
      if (res == FALSE) {
        # Try using pager2 instead
        pager2 <- getOption("pager2")
        if (is.null(pager2)) {
          stop("You must start Komodo Edit for displaying this file!")
        } else if (is.function(pager2)) {
          pager2(files = files, header = header, title = title,
            delete.file = delete.file)
        } else {
          file.show(files, header = header, title = title,
            delete.file = delete.file, pager = pager2)
        }
      } else {
        if (delete.file)
         koCmd(sprintf('window.setTimeout("try { sv.tools.file.getfile(\\"%s\\").remove(false); } catch(e) {}", 10000);', files));
      }
    }

  options(browser2 = getOption("browser"))
   svBrowser <- function(url) {
      url <- gsub("\\", "\\\\", url, fixed = TRUE)
      # If the URL starts with '/', I could safely assume a file path
      # on Unix or Mac and prepend 'file://'
      url <- sub("^/", "file:///", url)
      res <- tryCatch(koCmd(sprintf("sv.command.openHelp(\"%s\")", url)),
        error = function(e) return(FALSE))
     if (res == FALSE) {
        # Try using browser2 instead
        browser2 <- getOption("browser2")
        if (is.null(browser2)) {
          stop("You must start Komodo Edit for browsing files")
        } else {
          browseURL(url, browser = browser2)
        }
      }
    }

    options(editor2 = getOption("editor"))
    options(editor = Komodo, browser = svBrowser, pager = svPager)
  } else {
    Komodo <- NULL

    # TODO: use packageStartupMessage instead!
    #cat("R cannot find Komodo.", file = stderr())
    #if (.Platform$OS.type == "unix") {
    #	cat("Please, follow instructions at",
    #		"http://www.sciviews.org/SciViews-K to install it correctly.",
    #		"In particular, you must create a symbolic link in /user/local/bin:",
    #		"sudo ln -s <KomodoBinLocation>/komodo /usr/local/bin/komodo",
    #		"otherwise, R cannot find it!", sep = "\n", file = stderr())
    #} else {
    #	cat("Please, make sure you install it correctly\n",
    #		"You can find it at http://www.activestate.com/Products/komodo_edit.\n",
    #		file = stderr())
    #}
  }

  # Make sure we use HTML help (required for Alt-F1 and Alt-Shift-F1)
  # to display R help in Komodo Edit
  options(help_type = "html")
  # Make sure the help server is started
  port <- try(suppressMessages(tools::startDynamicHelp()), silent = TRUE)
  if (inherits(port, "try-error")) {
    # Dynamic help is already started
    if (R.Version()$`svn rev` >= 67550) {
      port <- try(suppressMessages(tools::startDynamicHelp(NA)), silent = TRUE)
    } else {
      port <- getNamespace("tools")$httpdPort
    }
  }
  # Record the home page for the help server in an option
  options(helphome = paste("http://127.0.0.1:", port,
    "/doc/html/index.html", sep = ""))

  # I need to get the help file URL, but help() does not provide it any
  # more! This is a temporary workaround for this problem
  assignTemp("getHelpURL", function(x, ...) {
    file <- as.character(x)
    if (length(file) == 0)
      return("")
    # Extension ".html" may be missing
    htmlfile <- basename(file)
    # Get the HTML help server port
    if (R.Version()$`svn rev` >= 67550) {
      port <- try(suppressMessages(tools::startDynamicHelp(NA)), silent = TRUE)
    } else {
      port <- getNamespace("tools")$httpdPort
    }
    if (length(file) > 1) {
      # If more then one topic is found
      paste("http://127.0.0.1:", port,
        "/library/NULL/help/", attr(x,"topic"), sep = "")
    } else {
      if (substring(htmlfile, nchar(htmlfile) - 4) != ".html")
        htmlfile <- paste(htmlfile, ".html", sep = "")
      paste("http://127.0.0.1:", port,
        "/library/", basename(dirname(dirname(file))),
        "/html/", htmlfile, sep = "")
    }
  })

# print() method of object returned by help() is very unflexible for R.app and
# does not allow in any way to use anything else than the R.app internal
# browser for help!!!
# That makes me very unhappy! Hey guys, I would like to use SciViews help
# browser here! So, no other solution than to be even harsher, and to force
# rewriting of the print function in base environment!!!
# (problem emailed to Simon Urbanek on 03/11/2009... I hope he will propose
# a work-around for this in R 2.12!!!)
#if (compareVersion(rVersion, "2.11.0") < 0) {
#	source("print.help_files_with_topic210.R")
#} else if (compareVersion(rVersion, "2.14.0") < 0) {
#	source("print.help_files_with_topic211.R")
#} else {
#	source("print.help_files_with_topic214.R")
#}

  R.initdir <- getOption("R.initdir")
  if (!is.null(R.initdir) && !is.na(R.initdir)) {
    # Change the working directory to the provided directory
    try(setwd(R.initdir), silent = TRUE)

    # Create a .Last.sys function that clears some variables in .GlobalEnv
    # and then, switch to R.initdir before closing R. The function is
    # stored in SciViews:TempEnv
    assignTemp(".Last.sys", function() {
      # Eliminate some known hidden variables from .GlobalEnv to prevent
      # saving them in the .RData file
      if (exists(".required", envir = .GlobalEnv, inherits = FALSE))
        rm(list = ".required", envir = .GlobalEnv, inherits = FALSE)
      # Note: .SciViewsReady is now recorded in SciViews:TempEnv instead
      # of .GlobalEnv, but we leave this code for old workspaces...
      if (exists(".SciViewsReady", envir = .GlobalEnv, inherits = FALSE))
        rm(list = ".SciViewsReady", envir = .GlobalEnv, inherits = FALSE)
      # If a R.initdir is defined, make sure to switch to it, so that
      # the session's workspace and command history are written at the
      # right place (in case of error, no change is made!)
      try(setwd(getOption("R.initdir")), silent = TRUE)
      # Clean up everything in Komodo
      tryCatch(svKomodo::koCmd("window.setTimeout(\"sv.r.closed();\", 1000);"),
        error = function(e) invisible(NULL))
    })

    msg <- paste("Session directory is", dQuote(R.initdir))
    msg2 <- NULL

    # Do we load .RData and .Rhistory now?
    args <- commandArgs()
    if (!"--vanilla" %in% args && !"--no-restore" %in% args &&
      !"--no.restore-data" %in% args) {
      if (file.exists(".RData")) {
        load(".RData", envir = .GlobalEnv)
        msg2 <- append(msg2, "data loaded")
      } else {
        msg2 <- append(msg2, "no data")
      }

      if (file.exists(".Rhistory")) {
        # On R Tk gui:
        # "Error in loadhistory(file) : no history mechanism available"
        # So, do it inside a try()
        history.loaded <- try(loadhistory(), silent = TRUE)
        if (inherits(history.loaded, "try-error"))  {
          msg2 <- append(msg2, "history cannot be loaded")
        } else {
          msg2 <- append(msg2, "history loaded")
        }
      } else {
        msg2 <- append(msg2, "no history")
      }
    } else {
      msg2 <- append(msg2, "data and history not loaded")
    }

    packageStartupMessage(paste(msg, " (", paste(msg2, collapse = ", "),
      ")", sep = ""))

    # Do we reactivate Komodo now?
    #	koact <- getOption("ko.activate")
    #	debugMsg("Reactivate Komodo:", koact)
  #	if (getTemp(".SciViewsReady", FALSE) && koact) {
  #		if ((.Platform$pkgType == "mac.binary")) {
  #			system("osascript -e 'tell application \"Komodo\" to activate'",
  #				wait = FALSE)
  #		} else if (!is.null(Komodo)) {
  #			## TODO: The following starts komodo if not started yet,
  #			## but does not activate it!
  #			system(shQuote(Komodo), wait = FALSE)
  #		}
  # Indicate to Komodo that R is ready
  # and test also communication from R to Komodo!
    #	koCmd('sv.cmdout.message("<<<data>>>", 10000, true);',
    #		data = paste("'", getOption("R.id"), "' (R ",
    #		R.Version()$major, ".", R.Version()$minor,
    #		") connected. Session dir: ",
    #		path.expand(getOption("R.initdir")), sep = ""))
    # ... and refresh the object explorer
    # TODO!

    # Differ synching R <-> Komodo to avoid deadlock situation
    # That does not work!
    #koCmd('window.setTimeout("sv.r.objects.getPackageList(true, true, true);", 500)')
    #koCmd('window.setTimeout("sv.r.test(true, true);", 500)')
  #	}
    # Update info in Komodo
    #debugMsg("Contacting Komodo with koCmd")
    #invisible(koCmd(paste(
    #	"sv.socket.rUpdate()",
    #	"sv.cmdout.append('R is started')",
    #	"sv.command.updateRStatus(true)",
    #	sep = ";"))
    #)
    # Refreshing Komodo's GUI elements
  #		try(koCmd(paste('sv.r.running = true; sv.socket.charset = "',  localeToCharset()[1],
  #			'"; sv.cmdout.message("' , R.version.string, ' is ready!");',
  #			' window.setTimeout("sv.r.objects.getPackageList(true, true, true);", 1000);',
  #			sep = "")), silent = TRUE)
    #try(koRefresh(force = TRUE), silent = TRUE)
    #}

    # TODO: eliminate this first from svStart.R... otherwise, it is run twice!
    # Do we have a .Rprofile file to source?
  #	rprofile <- file.path(c(getwd(), Sys.getenv("R_USER")), ".Rprofile")
  #	rprofile <- rprofile[file.exists(rprofile)][1]
  #	if (!is.na(rprofile)) {
  #		source(rprofile)
  #		debugMsg("Loaded file:", rprofile)
  #	}

    # If there is a function called ".First" in .GlobalEnv, run it now
    .First <- NULL
    if (exists(".First", envir = .GlobalEnv, mode = "function",
      inherits = FALSE))
      eval(.First(), envir = .GlobalEnv)
  }
}

.onUnload <- function(libpath) {
  # Restore R options that preexisted
  try(options(useFancyQuotes = getTemp(".useFancyQuotes", TRUE)),
    silent = TRUE)
  try(options(max.print = getTemp(".max.print", 99999)),
    silent = TRUE)

  # Restore editor, browser & pager
  editor2 <- getOption("editor2")
  if (!is.null(editor2)) {
    options(editor = editor2)
    options(editor2 = NULL)
  }
  browser2 <- getOption("browser2")
  if (!is.null(browser2)) {
    options(browser = browser2)
    options(browser2 = NULL)
  }
  pager2 <- getOption("pager2")
  if (!is.null(pager2)) {
    options(pager = pager2)
    options(pager2 = NULL)
  }
  #serve <- getOption("ko.serve")
  #if (!is.null(serve) && "package:svSocket" %in% search())
  #  stopSocketServer(port = as.integer(serve)[1])
  koUninstall()
  # Remove the SciViews task callback manager
  try(removeTaskCallback("SV-taskCallbackManager"), silent = TRUE)
  try(rmTemp(".svTaskCallbackManager"), silent = TRUE)

  # Restore rc.settings
  settings <- getTemp(".old.rc.settings", rc.settings())
  do.call("rc.settings", as.list(settings))
  rmTemp(".old.rc.settings")
}
